### ISSUE TITLE
Invoice issue - Enhancement Requests

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1320]
In GitLab by @kdjstudios on Jan 10, 2019, 13:27

**Submitted by:** Support <support@sabilling.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2019-01-09-37292

**Server:** Internal (All)

**Client/Site:** All

**Account:** All

**Issue:** 

I just wanted to follow up and recap our phone call. We have initial changes proposed to the current invoice labels and calculations and we also have overall changes to the invoice structure, each separated out below.

Invoice Labels and Calculations:
1. Update "Total" label to something along the lines of the following: "Account Balance"; "Ending Balance"; or "Current Account Balance". This will better portray that this value is the current/total account balance.
2. Update the calculation for "New charges"; as currently 'over payments' will effect this and can lead to confusion. 
 - This new calculation will only include current invoice billable items; but will not include any payments.
 - Or update the label to "New Charges and Payments"; this seems a bit less useful then the above however. As most clients would like the see the total for just the new charges. 

Invoice Structure:
1. We would like to update the invoice structure along with the changes above. I have drafted an attached mock up of just the 'invoice' items to hopefully better display it. Rich your feedback and suggestions would be greatly appreciated on this.
 - Basically we are wanting to separate out each line item between payments and billing items. Then add subtotals to each. 

![Invoice_Mock_Up](/uploads/711e31cbe1618c8279175bcf5af6c83d/Invoice_Mock_Up.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 10, 2019, 13:33

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 10, 2019, 13:34

@smasih @pchaudhary @amishra 

May we please get an estimate on the first part:

- Update "Total" label to something along the lines of the following: "Account Balance"; "Ending Balance"; or "Current Account Balance". This will better portray that this value is the current/total account balance.

- Update the calculation for "New charges"; as currently 'over payments' will effect this and can lead to confusion. 

    - This new calculation will only include current invoice billable items; but will not include any payments.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 19, 2019, 11:50

Updated priority.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 24, 2019, 07:08

@kdjstudios 

Thank you for the update.

|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|4|Aware of Concepts, implementation tricky|7|High Impact|5.5|12.5|

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 5, 2019, 09:57

@amishra @tim.traylor 

I think we have had other invoice layout changes in the past three months since this was originally requested and roughly 13 hours for an updated layout seems very high. I am going to have to wait for @tim.traylor to approve of this. 

Also, I think, would this be better as an "new" invoice format? I think each site has the drop down to choose which invoice format it uses correct?

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 5, 2019, 12:26

@amishra @kdjstudios Kyle, I believe this would replace the Answernet format.  We could also check with our customers and see if they'd prefer the new format.

And I agree that 12.5 hours is a lot. Ashish, can you pls run that by Sumeet?  Thx

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 6, 2019, 02:30

@kdjstudios @tim.traylor 



> And I agree that 12.5 hours is a lot. Ashish, can you pls run that by Sumeet?  Thx

Yes, I understand 12.5 is high for format change, but it's not js/css/html its a ruby library and we need to follow their syntax and there is not much flexibility they provide.

A little change can lead to incorrect demarcation of the whole invoice. 

@pchaudhary @amishra 

Please go through the requirements again and re-estimate because I think it might take more than 12.5.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Aug 19, 2019, 05:32

@kdjstudios @tim.traylor 

Thank you for the update.

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|4|R&D Required|10|High Impact|6.0|16|

##### Estimate Notes:
* Risk areas identified: `Invoice/Billing`
* Work plan: 
   * As per the current functionality, each accounts line items subtotal comes at the end and then payment, etc...
  * We are using the ruby library for download pdf and there is not much flexibility provided and even there is no proper documentation for that. So, we need to do R & D.
  * Due to which there may be multiple retries for the fix and testing.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 20, 2019, 24:33

added 1h 3m of time spent

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 20, 2019, 24:33

added 30m of time spent

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 30, 2019, 07:47

@smasih @amishra @kdjstudios @jredfield Sumeet and Asish, There is a 107 page manual for Prawn, so I don't know why it was commented that there is no proper documentation.  

John, I'm not comfortable approving 16 hours to add 3 sub-total lines to invoices.  I expect you know more about this than I do, so can you pls address the approval?  Thx

[manual.pdf](/uploads/42d7adfe09c8417f6037313757cb38fe/manual.pdf)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 30, 2019, 09:47

noted thanks Tim

@pchaudhary @amishra  let's discuss after higher priority items are completed.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 17, 2019, 03:17

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 17, 2019, 03:25

@kdjstudios @jredfield @tim.traylor 

Thank you for the update.

* I understand, but we need to change/modified the old pdf design, It's a ruby/gem library change not in html/css file and we need to follow their syntax and there is not much flexibility they provide.

* A little change can lead to incorrect demarcation of the whole invoice.

**NOTE: I will try to complete under the given estimate..**

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 17, 2019, 10:59

As previously stated by Tim, 16 hours to do this is not approved.

marking this for @jredfield to review and determine if he is able to do this in less time.
